File: 82-remove-duplicates-from-sorted-list-ii/82-remove-duplicates-from-sorted-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

/// Iteration over recursion
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null){
            return null;
        }
        
        ListNode dummy = new ListNode(-1); 
        dummy.next = head;
        ListNode curr = head;
        ListNode store = dummy;
        
        while (curr != null){
            while (curr.next != null && curr.next.val == curr.val){
                curr = curr.next;
            }
            
            if (store.next == curr){
                store = curr;
            }else{
                store.next = curr.next;
            }
            
            curr = curr.next;
            
        }
        
        return dummy.next;
    }
}

File: 82-remove-duplicates-from-sorted-list-ii/82-remove-duplicates-from-sorted-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        if head == None:
            return None
        
        trav = head
        dup = False
        
        while trav.next and trav.next.val == trav.val:
            trav = trav.next
            dup = True
            
        if dup:
            trav = self.deleteDuplicates(trav.next)
        else:
            trav.next = self.deleteDuplicates(trav.next)
        
        return trav
            